lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp

```diff
@@ -0,0 +1,318 @@
+        return op->emitError("Supplied value of scalar constant exceeds limits "
+                             "of destination type");
+      }
+      int64_t d = (isFloat ? static_cast<int64_t>(doubleValue) : intValue);
+      if (doBroadcast) {
+        mhloTensor = getSplatConstTensor<int64_t>(rewriter, op, d, dtype, dshape);
+      } else {
+        mhloTensor =
+            mhlo::getConstTensor<int64_t>(rewriter, op, {d}, dshape).getValue();
+      }
+    }
+  } else
+    return op->emitError("Usupported element type");
+
+  return success();
+}
+
+
+LogicalResult torchAlphaToMhloTensor(ConversionPatternRewriter &rewriter,
+                                     Operation *op, Value alphaScalar,
+                                     Value &alphaTensor, Type dtype,
+                                     llvm::ArrayRef<int64_t> dshape,
+                                     bool checkForUnity) {
+  if (succeeded(torchScalarToMhloTensor(rewriter, op, alphaScalar, alphaTensor,
+                                        dtype, dshape)))
+    return success();
+
+  // `alpha` has not been specified.
+  int64_t alphaValue;
+  if (!matchPattern(alphaScalar, m_TorchConstantInt(&alphaValue)))
+    return op->emitError("Currently only scalar constants are supported for "
+                         "alpha in MHLO operation");
+  // When no alpha has been specified, this must be 1.
+  if (checkForUnity && alphaValue != 1)
+    return op->emitError("Unsupported integer value for alpha");
+
+  alphaTensor =
+      mlir::mhlo::getMhloConstTensorSingleF32(rewriter, op, alphaValue);
+
+  return success();
+}
+
+
+Value promoteAndBroadcast(ConversionPatternRewriter &rewriter,
+                          Value input, TensorType outType) {
+  // Two tensors are “broadcastable” if the following rules hold:
+  //   - Each tensor has at least one dimension.
+  //   - When iterating over the dimension sizes, starting at the trailing dimension,
+  //   the dimension sizes must either be equal, one of them is 1, or one of them
+  //   does not exist.
+  Operation* op = input.getDefiningOp();
+  TensorType in_type = input.getType().dyn_cast<TensorType>();
+
+  if (in_type.getElementType() != outType.getElementType()) {
+    TensorType promoted_type = in_type.cloneWith(in_type.getShape(), outType.getElementType());
+    input = rewriter.create<mhlo::ConvertOp>(op->getLoc(), promoted_type, input);
+  }
+
+  ArrayRef<int64_t> inShape = in_type.getShape();
+  ArrayRef<int64_t> outShape = outType.getShape();
+
+  bool do_bcast = (inShape.size() != outShape.size());
+  SmallVector<int64_t> bcastDims;
+  for (size_t i = 0; i < inShape.size(); ++i) {
+    // iterating over the dimension sizes, starting at the trailing dimension
+    size_t outPos = outShape.size() - 1 - i;
+    size_t inPos = inShape.size() - 1 - i;
+    int64_t outDim = outShape[outPos];
+    int64_t inDim = inShape[inPos];
+    if (inDim == outDim) {
```

Comment:
The implementation doesn't support implicit broadcast when `inDim == outDim == -1` both unknown?